crates/sdk/src/lib.rs

```diff
@@ -273,6 +302,152 @@ impl<E: StarkFriEngine<SC>> GenericSdk<E> {
+        );
+
+        // Fill out the public values length and OpenVM version in the template
+        let openvm_verifier_code = OPENVM_HALO2_VERIFIER_TEMPLATE
+            .replace("{PUBLIC_VALUES_LENGTH}", &pvs_length.to_string())
+            .replace("{OPENVM_VERSION}", env!("CARGO_PKG_VERSION"));
+
+        // Create temp dir
+        let temp_dir = tempdir().wrap_err("Failed to create temp dir")?;
+        let temp_path = temp_dir.path();
+
+        // Make interfaces dir
+        let interfaces_path = temp_path.join("interfaces");
+        create_dir_all(&interfaces_path)?;
+
+        // Write the files to the temp dir. This is only for compilation
+        // purposes.
+        write(
+            interfaces_path.join(OPENVM_HALO2_VERIFIER_INTERFACE_NAME),
+            OPENVM_HALO2_VERIFIER_INTERFACE,
+        )?;
+        write(
+            temp_path.join(OPENVM_HALO2_VERIFIER_PARENT_NAME),
+            &halo2_verifier_code,
+        )?;
+        write(
+            temp_path.join(OPENVM_HALO2_VERIFIER_BASE_NAME),
+            &openvm_verifier_code,
+        )?;
+
+        // Run solc from the temp dir
+        let output = Command::new("solc")
+            .current_dir(temp_path)
+            .arg("OpenVmHalo2Verifier.sol")
+            .arg("--no-optimize-yul")
+            .arg("--bin")
+            .arg("--optimize")
+            .arg("--optimize-runs")
+            .arg("100000")
+            .output()?;
+
+        if !output.status.success() {
+            eyre::bail!(
+                "solc exited with status {}: {}",
+                output.status,
+                String::from_utf8_lossy(&output.stderr)
+            );
+        }
+
+        let bytecode = Self::extract_binary(
+            &output.stdout,
+            "OpenVmHalo2Verifier.sol:OpenVmHalo2Verifier",
+        );
+
+        let evm_verifier = OpenVmEvmHalo2Verifier {
+            halo2_verifier_code,
+            openvm_verifier_code,
+            openvm_verifier_interface: OPENVM_HALO2_VERIFIER_INTERFACE.to_string(),
+            artifact: EvmVerifierByteCode {
+                sol_compiler_version: "0.8.19".to_string(),
+                sol_compiler_options: "--no-optimize-yul --bin --optimize --optimize-runs 100000"
+                    .to_string(),
+                bytecode,
+            },
+        };
+        Ok(evm_verifier)
+    }
+
+    /// We will split the output by whitespace and look for the following
+    /// sequence:
+    /// [
+    ///     ...
+    ///     "=======",
+    ///     "OpenVmHalo2Verifier.sol:OpenVmHalo2Verifier",
+    ///     "=======",
+    ///     "Binary:"
+    ///     "[compiled bytecode]"
+    ///     ...
+    /// ]
+    ///
+    /// Once we find "OpenVmHalo2Verifier.sol:OpenVmHalo2Verifier," we can skip
+    /// to the appropriate offset to get the compiled bytecode.
+    fn extract_binary(output: &[u8], contract_name: &str) -> Vec<u8> {
```

Comment:
these are internal functions just for EVM verify; make a separate module `evm` for all the evm types and methods not directly callable from `Sdk`